### PR TITLE
refactor: support webpack 5

### LINKED ISSUE #223: Supporting Webpack 5
Testing application with Webpack 5.0.0
Warning emitted
```
warning: worker-loader@2.0.0" has incorrect peer dependency "webpack@^3.0.0 || ^4.0.0-alpha.0 || ^4.0.0".
```

Would be good to have peerDependencies set to
"webpack": " ^4.0.0 || ^5.0.0"

https://github.com/webpack/changelog-v5/blob/master/MIGRATION%20GUIDE.md

### PR DESCRIPTION
**Pull request by smelukov**
This PR contains a:

- [ ] **bugfix**
- [ ] new **feature**
- [x] **code refactor**
- [x] **test update** 
- [ ] **typo fix**
- [ ] **metadata update**

### Motivation / Use-Case

Support webpack 5 (closes #223)

### Breaking Changes

No

### Additional Info

This PR blocks this [one](https://github.com/webpack/webpack/pull/9975)

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -70,45 +71,58 @@ export function pitch(request) {
...
-      if (options.fallback === false) {
-        delete this._compilation.assets[worker.file];
-      }
-
-      return cb(
-        null,
-        `module.exports = function() {\n  return ${worker.factory};\n};`
+      worker.file = [...entries[0].files][0];
+
+      const cacheIdent = `${worker.compiler.compilerPath}${__dirname}/${
```
Comment (sokra) [inline src/index.js]: Should probably include `worker-loader`

File: src/index.js [hunk 2]
```diff
@@ -70,45 +71,58 @@ export function pitch(request) {
...
+            return cb(null, content);
+          }
+
+          worker.factory = getWorker(
+            worker.file,
+            compilation.assets[worker.file].source(),
+            options
+          );
+
+          if (options.fallback === false) {
```
Comment (sokra) [inline src/index.js]: This also need to happen in the case where the content is cached

File: package.json [hunk 3]
```diff
@@ -39,7 +39,7 @@
     "dist"
   ],
   "peerDependencies": {
-    "webpack": "^3.0.0 || ^4.0.0-alpha.0 || ^4.0.0"
+    "webpack": "^5.0.0-beta.5"
```
Comment (evilebottnawi) [inline package.json]: We should support webpack@4 too

Comment (smelukov) [reply package.json]: This PR is about dropping webpack 4 support, and I expect that these changes will be published under major version update ;)

Comment (evilebottnawi) [reply package.json]: @smelukov better keep compatibility with webpack@4 and webpack@5 some times

Comment (evilebottnawi) [reply package.json]: Also `^5.0.0-beta.6` not working with `^5.0.0-beta.5`

Comment (smelukov) [reply package.json]: > Also ^5.0.0-beta.6 not working with ^5.0.0-beta.5

Not sure that understood you correctly...
I've used `^5.0.0-beta.5` for now and it correctly matched for `5.0.0-beta.6` and below
https://jubianchi.github.io/semver-check/#/^5.0.0-beta.5/5.0.0-beta.6

Comment (evilebottnawi) [reply package.json]: Just put here `^5.0.0` and ignore npm warning

File: package.json [hunk 4]
```diff
@@ -70,7 +70,7 @@
     "pre-commit": "^1.0.0",
     "prettier": "^1.13.4",
     "standard-version": "^4.0.0",
-    "webpack": "^3.0.0"
+    "webpack": "^5.0.0-beta.5"
```
Comment (evilebottnawi) [inline package.json]: Use `next`

File: package.json (hunk 3 above)
Comment (evilebottnawi) [reply package.json]: Anyway i still think we should support webpack@4 and webpack@5

Comment (smelukov) [reply package.json]: Support webpack 4 as well ;)

File: package.json [hunk 5]
```diff
@@ -39,7 +39,7 @@
     "dist"
   ],
   "peerDependencies": {
-    "webpack": "^3.0.0 || ^4.0.0-alpha.0 || ^4.0.0"
+    "webpack": "^3.0.0 || ^4.0.0 || ^5.0.0"
```
Comment (evilebottnawi) [inline package.json]: Let's drop `webpack@3` from supporting

**Comment by smelukov** [comment]
@evilebottnawi done with it )

**Comment by colelawrence** [comment]
@sokra, would you be able to review this or delegate?

**Comment by paolobarbolini** [comment]
@sokra Friendly ping as this fixes the last issue I have before I can upgrade to Webpack 5 😃

File: package.json [hunk 6]
```diff
@@ -70,7 +70,7 @@
     "pre-commit": "^1.0.0",
     "prettier": "^1.13.4",
     "standard-version": "^4.0.0",
-    "webpack": "^3.0.0"
+    "webpack": "next"
```
Comment (sokra) [inline package.json:73]: Should be something like `^5.0.0-beta.15`

File: src/supportWebpack5.js [hunk 7]
```diff
@@ -0,0 +1,62 @@
+import getLazyHashedEtag from 'webpack/lib/cache/getLazyHashedEtag';
```
Comment (sokra) [inline src/supportWebpack5.js:1]: This need to be exposed in a better way in webpack, but fine for now

File: package.json (hunk 6 above)
Comment (evilebottnawi) [reply package.json:73]: I will fix it in other PRs

**Comment by Bnaya** [comment]
Is this released somewhere?

**Comment by evilebottnawi** [comment]
ETA is near future

**Comment by smelukov** [comment]
Dont forget to review https://github.com/webpack/webpack/pull/9975

**Comment by Bnaya** [comment]
It will be great to have it as @next or so :)

**Comment by Birch-san** [comment]
awaiting this release also! :) it's the final thing we require in order to move to webpack 5.

**Comment by evilebottnawi** [comment]
@Birch-san be patient we want to solve more issues in the next version

**Pull request merged**